### PR TITLE
Sort compound checker diagnostic messages.

### PR DESCRIPTION
**Pull request by smillst**
Gather all diagnostic messages (e.g., error messages) for a compilation unit.  Sort them by source code position and then by checker (subcheckers first).  Print them in this order.

This only gathers messages that use a tree as a source positions.  Other messages, such as calls to ErrorReporter.errorAbort, are issued immediately.

### CODE REVIEW

File: checker/jtreg/sortwarnings/OrderOfCheckers.out [hunk 1]
```diff
@@ -0,0 +1,19 @@
+OrderOfCheckers.java:15:60: compiler.err.proc.messager: [[all, value]:assignment.type.incompatible] incompatible types in assignment.
+found   : @UnknownVal int @UnknownVal []
```
Comment (wmdietl) [inline checker/jtreg/sortwarnings/OrderOfCheckers.out]: Do you also want to use `-Anomsgtext` to make the output here look like in `ErrorOrders.out`? All the details don't seem necessary.
If the distinction is important, can you document why?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 2]
```diff
@@ -118,23 +125,29 @@ public void initChecker() {
...
+     * of its subcheckers both depend on a third checker, that checker will only be instantiated
+     * once.
      *
-     * <p>This method is protected so it can be overridden, but it is only intended to be called
-     * internally by the BaseTypeChecker. Please override this method but do not call it from
-     * classes other than BaseTypeChecker. Subclasses that override this method should call super
-     * and added dependencies so that checkers required for reflection resolution are included if
-     * reflection resolution is requested.
+     * <p>Though each checker is run on a whole compilation unit before the next checker is run,
+     * error and warning messages collected and sorted based on location in the source file before
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: collected -> are collected?
on location -> on their location?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 3]
```diff
@@ -118,23 +125,29 @@ public void initChecker() {
...
+     * being printed. (See {@link #printMessage(Kind, String, Tree, CompilationUnitTree)}.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: Closing `)` for `(See` missing.

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 4]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+            printCollectedMessages(tree.getCompilationUnit());
+            this.errsOnLastExit = log.nerrors;
+        }
+    }
+
+    /**
+     * Stores all messages issued by this checker and its subcheckers for the current compilation
+     * unit. The messages are printed after all checkers have processed the current compilation
+     * unit. If this checker has no subcheckers and is not a subchecker for any other checker, then
+     * messageStore is null and messages will printed as they are issued by this checker.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: will printed -> will be printed.

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 5]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+     */
+    private TreeSet<CheckerMessage> messageStore = null;
+
+    /**
+     * If this is a compound checker or a subchecker of a compound checker, then the message is
+     * stored until all message from all checkers for the compilation unit are issued.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: all message -> all messages

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 6]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+    /**
+     * Prints error messages for this checker and all subcheckers such that the errors are ordered
+     * by line and column number and then by checker. (See checkerMessageComparator for more precise
+     * order.)
+     *
+     * @param unit Current compilation unit
+     */
+    private void printCollectedMessages(CompilationUnitTree unit) {
+        if (messageStore != null) {
+            for (CheckerMessage pos : messageStore) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: `pos` isn't really a good name for this variable. How about `message` or `msg`?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 7]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
             int errorsAfterTypeChecking = log.nerrors;
             nerrorsOfAllPreviousCheckers += errorsAfterTypeChecking - errorsBeforeTypeChecking;
         }
+
         this.errsOnLastExit = nerrorsOfAllPreviousCheckers;
         super.typeProcess(element, tree);
+
+        if (getSubcheckers().size() > 0) {
+            printCollectedMessages(tree.getCompilationUnit());
+            this.errsOnLastExit = log.nerrors;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: Why is this change needed? Maybe some adapted version of the removed comments (lines 426/430) would help explain?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 8]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+            if (kind != that.kind) {
+                return false;
+            }
+            if (!message.equals(that.message)) {
+                return false;
+            }
+            if (!source.equals(that.source)) {
+                return false;
+            }
+            return checker.equals(that.checker);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: I think neither Tree nor Checker implement equals, so you could just use reference equality.

File: framework/src/org/checkerframework/framework/source/SourceChecker.java [hunk 9]
```diff
@@ -1827,7 +1837,8 @@ public final String getOption(String name, String def) {
     protected final Collection<String> getStandardSuppressWarningsKeys() {
         SuppressWarningsKeys annotation = this.getClass().getAnnotation(SuppressWarningsKeys.class);
 
-        Set<String> result = new HashSet<>();
+        // This ensures keys are printed consistently
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/source/SourceChecker.java]: `TreeSet ensures keys are returned in a consistent order`

File: checker/jtreg/sortwarnings/ErrorOrders.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package index;
+
+import org.checkerframework.checker.index.qual.GTENegativeOne;
+import org.checkerframework.checker.index.qual.IndexFor;
+import org.checkerframework.checker.index.qual.MinLenBottom;
+import org.checkerframework.checker.index.qual.SameLenBottom;
+import org.checkerframework.checker.index.qual.UpperBoundBottom;
+import org.checkerframework.common.value.qual.BottomVal;
+
+/** This class tests that error are issued in order of postion. */
```
Comment (wmdietl) [inline checker/jtreg/sortwarnings/ErrorOrders.java]: error -> errors

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 11]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+            new Comparator<CheckerMessage>() {
+
+                @Override
+                public int compare(CheckerMessage o1, CheckerMessage o2) {
+                    int byPos = InternalUtils.compareDiagnosticPosition(o1.source, o2.source);
+                    if (byPos != 0) {
+                        return byPos;
+                    }
+
+                    int o1Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o1.checker);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: It's a bit hard to read that `o1.checker` and `o1Checker` have different types. How about `o1CheckerIndex`?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 12]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+                        return kind;
+                    }
+
+                    return o1.message.compareTo(o2.message);
+                }
+            };
+
+    /** Represents a message (e.g., an error message) issued by a checker. */
+    private class CheckerMessage {
+        Diagnostic.Kind kind;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: I would at least add a `final` to all the fields.

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 13]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+                    if (kind != 0) {
+                        return kind;
+                    }
+
+                    return o1.message.compareTo(o2.message);
+                }
+            };
+
+    /** Represents a message (e.g., an error message) issued by a checker. */
+    private class CheckerMessage {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: Can this class be static - which would safe us from storing a reference to the enclosing class?

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 14]
```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+
+                @Override
+                public int compare(CheckerMessage o1, CheckerMessage o2) {
+                    int byPos = InternalUtils.compareDiagnosticPosition(o1.source, o2.source);
+                    if (byPos != 0) {
+                        return byPos;
+                    }
+
+                    int o1Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o1.checker);
+                    int o2Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o2.checker);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: How about computing and storing this index with the CheckerMessage, either in addition or instead of `checker`?

Comment (smillst) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: A subchecker can't access the list of other subcheckers, so it can't compute this number and therefore can't create a CheckerMessage with this index.  I could add a field to BaseTypeChecker to hold this number.

File: checker/jtreg/sortwarnings/OrderOfCheckers.out (hunk 1 above)
Comment (smillst) [inline checker/jtreg/sortwarnings/OrderOfCheckers.out]: I want to use `-AshowSuppressWarningKeys` so that this tests which checker is issuing the error. If I use `-Anomsgtext`, the keys aren't printed.  I documented this in Main.java.

File: checker/jtreg/sortwarnings/Main.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+/*
+ * @test
+ * @summary Test that errors from compound checkers are sorted correctly.
+ * The first compilation below tests that the errors are ordered by line/column number.  The second line tests that the errors at the same line/column are sorted in order of checker; hence, -Anomsgtext is not passed so that -AshowSuppressWarningKeys has an effect.
```
Comment (wmdietl) [inline checker/jtreg/sortwarnings/Main.java]: Wrap very long line.

File: framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 16]
```diff
@@ -412,35 +425,182 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
...
+            new Comparator<CheckerMessage>() {
+
+                @Override
+                public int compare(CheckerMessage o1, CheckerMessage o2) {
+                    int byPos = InternalUtils.compareDiagnosticPosition(o1.source, o2.source);
+                    if (byPos != 0) {
+                        return byPos;
+                    }
+
+                    int o1Index = BaseTypeChecker.this.getSubcheckers().indexOf(o1.checker);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: Please add some documentation to `CheckerMessage.checker` and this computation to make the relationship between the different checkers clearer.
Also document that the `checkerMessageComparator` is always only used when the enclosing `BaseTypeChecker.this` is a compound checker (or only if it is the top-most checker?).

Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java]: Or should this code be changed to leave `checkerMessageComparator` null, if this is not a compound checker?
